[Export] Test non-strict mode on existing test cases

andrewlee302 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115413
* #115402
* __->__ #115399

Summary:
Dynamo test methodology provides a good example to patch various
treaments on the same set of test cases. A pitfall is the global config
that could be easily modified somewhere. Here we change the behavior of
the export API thru hijacking it with self defined code.

For supporting non-strict test suite, the `strict=False` is explicitly
passed into the export API when it's called w/ or w/o strict arg.

Test Plan:
python test/export/test_export_nonstrict.py

andrewlee302 committed:
[Export] Test non-strict mode on existing test cases

Summary:
Dynamo test methodology provides a good example to patch various
treaments on the same set of test cases. A pitfall is the global config
that could be easily modified somewhere. Here we change the behavior of
the export API thru hijacking it with self defined code.

For supporting non-strict test suite, the `strict=False` is explicitly
passed into the export API when it's called w/ or w/o strict arg.

* For existing failed strict test cases, non-strict also fails.
* For passed strict but failed non-strict cases, we mark them as
`@testing.expectedFailureNonStrict`.
* Moreover, I manually check the failure reason and some of them are not
related to nn.Module asserting exception. I mark them as `# Need to fix
for non-strict mode`.

Test Plan:
python test/export/test_export_nonstrict.py

[ghstack-poisoned]

andrewlee302 committed:
Update on "[Export] Test non-strict mode on existing test cases"


Summary:
Dynamo test methodology provides a good example to patch various
treaments on the same set of test cases. A pitfall is the global config
that could be easily modified somewhere. Here we change the behavior of
the export API thru hijacking it with self defined code.

For supporting non-strict test suite, the `strict=False` is explicitly
passed into the export API when it's called w/ or w/o strict arg.

* For existing failed strict test cases, non-strict also fails.
* For passed strict but failed non-strict cases, we mark them as
`testing.expectedFailureNonStrict`.
* Moreover, I manually check the failure reason and some of them are not
related to nn.Module asserting exception. I mark them as `# Need to fix
for non-strict mode`.

Test Plan:
python test/export/test_export_nonstrict.py

[ghstack-poisoned]

andrewlee302 commented on test/export/testing.py:
```diff
@@ -0,0 +1,43 @@
+            if not callable(fn):
+                setattr(DummyTestClass, name, getattr(cls, name))
+                continue
+            new_name = f"{name}{fn_suffix}"
```
You're right. E.g. `python test/export/test_export_nonstrict.py -k test_basic_non_strict` run the `TestExport.test_basic` with non-strict mode.

tugsbayasgalan commented on test/export/testing.py:
```diff
@@ -0,0 +1,43 @@
+    cls, cls_prefix, fn_suffix, mocked_export_fn, xfail_prop=None
+):
+    DummyTestClass = type(f"{cls_prefix}{cls.__name__}", cls.__bases__, {})
+    DummyTestClass.__qualname__ = DummyTestClass.__name__
```
nit: maybe name it MockedTestClass or sth?

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -40,10 +40,15 @@
     treespec_dumps,
     treespec_loads,
 )
+try:
```
Hmm..

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -90,6 +95,7 @@ def conflicting_constraints(x):
         ):
             ep(torch.tensor([3]))
 
+    @testing.expectedFailureNonStrict
```
It seems bad that we are leaking non-strict info into this test file, could we maybe track non-strict failures in the non_strict test file instead? I imagine once we have seperate files for "retrace", "ser_der", we can't be applying decorators on this file. One way you can do is you can create a list in "test_export_non_strict" file and make sure that each entry in that list fails.

andrewlee302 replied:
> I imagine once we have seperate files for "retrace", "ser_der", we can't be applying decorators on this file.

Actually, we can stack the decorators and it's what I do in the PR stack. 

> One way you can do is you can create a list in "test_export_non_strict" file and make sure that each entry in that list fails.

That's another approach and seems good in this PR. 

I regard a test unit as a typical usage from user's perspective. If we consider all the test variations (non-strict, retrace, ser_der), centralized marks (decorators) close to the test case could clearly show the view of support power for each test unit. If we scatter those marks into each variation, it's hard to identify the existing problems of the each test unit.

tugsbayasgalan replied:
> > I imagine once we have seperate files for "retrace", "ser_der", we can't be applying decorators on this file.
> 
> Actually, we can stack the decorators and it's what I do in the PR stack.
> 
> > One way you can do is you can create a list in "test_export_non_strict" file and make sure that each entry in that list fails.
> 
> That's another approach and seems good in this PR.
> 
> I regard a test unit as a typical usage from user's perspective. If we consider all the test variations (non-strict, retrace, ser_der), centralized marks (decorators) close to the test case could clearly show the view of support power for each test unit. If we scatter those marks into each variation, it's hard to identify the existing problems of the each test unit.



> Actually, we can stack the decorators and it's what I do in the PR stack.

Oh I meant from UX perspective.

tugsbayasgalan replied:
> I regard a test unit as a typical usage from user's perspective. If we consider all the test variations (non-strict, retrace, ser_der), centralized marks (decorators) close to the test case could clearly show the view of support power for each test unit. If we scatter those marks into each variation, it's hard to identify the existing problems of the each test unit.

I see, that's fair. @zhxchen17 @angelayi  what do u guys think?

tugsbayasgalan commented:
@andrewlee302 it might be also helpful if you can generate a google sheet listing all errors that was thrown while you were automatically generating test cases. That way, we can check if there are some systemic fixes we need to do.

andrewlee302 commented:
> @andrewlee302 it might be also helpful if you can generate a google sheet listing all errors that was thrown while you were automatically generating test cases. That way, we can check if there are some systemic fixes we need to do.

Agree that we need to fix them. For example, the non-strict test failures mostly come from the assertion that the callable must be a nn.Module. I think the more programmatic way is to remove `_expected_failure_non_strict` in the test and then run it.

andrewlee302 committed:
Update on "[Export] Test non-strict mode on existing test cases"


Summary:
Dynamo test methodology provides a good example to patch various
treaments on the same set of test cases. A pitfall is the global config
that could be easily modified somewhere. Here we change the behavior of
the export API thru hijacking it with self defined code.

For supporting non-strict test suite, the `strict=False` is explicitly
passed into the export API when it's called w/ or w/o strict arg.

* For existing failed strict test cases, non-strict also fails.
* For passed strict but failed non-strict cases, we mark them as
`testing.expectedFailureNonStrict`.
* Moreover, I manually check the failure reason and some of them are not
related to nn.Module asserting exception. I mark them as `# Need to fix
for non-strict mode`.

Test Plan:
python test/export/test_export_nonstrict.py

[ghstack-poisoned]

zhxchen17 reviewed (approved):
lgtm

tugsbayasgalan commented on test/export/test_export_nonstrict.py:
```diff
@@ -0,0 +1,51 @@
+from torch.export import export
+
+try:
+    from . import test_export, testing
```
Why do we have try/catch here?

andrewlee302 replied:
I copied the coding style from https://github.com/pytorch/pytorch/blob/4471fe6c3915f4afea918e1b3af95e7298ad42b4/test/dynamo/test_dynamic_shapes.py#L10-L31. Do you know why? Maybe it's related to being imported by other files. 

I think we can remove it for our purpose here.

tugsbayasgalan replied:
Yeah let's remove it. Seems weird

andrewlee302 commented on test/export/testing.py:
```diff
@@ -0,0 +1,43 @@
+def _make_fn_with_mocked_export(fn, mocked_export_fn):
+    @functools.wraps(fn)
+    def _fn(*args, **kwargs):
+        with patch("test_export.export", mocked_export_fn):
```
That's on purpose. Here we keep the original handler `torch.export.export` and only change the behavior of `test_export.export`. It won't break as long as test_export.py has `from torch.export import export`.

tugsbayasgalan replied:
Why don't we just patch the original handler? Since this is within this context manager, i feel like it should be fine for whoever is using export outside of this file

andrewlee302 replied:
The reason is we could introduce infinite recursion if we patch directly on the real export API. In the way here, we only patch the surfaced export in the test file. Does it make sense?

tugsbayasgalan replied:
Yeah makes sense. Can you add a comment on test_export.py that we shouldn't change the import statement?

andrewlee302 commented on test/export/test_export_nonstrict.py:
```diff
@@ -0,0 +1,51 @@
+    return export(*args, **kwargs, strict=False)
+
+
+def make_dynamic_cls(cls):
```
I think it's not necessary as it's only used privately in this file. The file has already indicated the non-strict context.

tugsbayasgalan replied:
Ok maybe leave a docstring describing what this is doing? Dynamic is pretty overloaded term in pytorch, so it might cause some confusion.

andrewlee302 committed:
Update on "[Export] Test non-strict mode on existing test cases"


Summary:
Dynamo test methodology provides a good example to patch various
treaments on the same set of test cases. A pitfall is the global config
that could be easily modified somewhere. Here we change the behavior of
the export API thru hijacking it with self defined code.

For supporting non-strict test suite, the `strict=False` is explicitly
passed into the export API when it's called w/ or w/o strict arg.

* For existing failed strict test cases, non-strict also fails.
* For passed strict but failed non-strict cases, we mark them as
`testing.expectedFailureNonStrict`.
* Moreover, I manually check the failure reason and some of them are not
related to nn.Module asserting exception. I mark them as `# Need to fix
for non-strict mode`.

Test Plan:
python test/export/test_export_nonstrict.py

[ghstack-poisoned]

andrewlee302 committed:
Update on "[Export] Test non-strict mode on existing test cases"


Summary:
Dynamo test methodology provides a good example to patch various
treaments on the same set of test cases. A pitfall is the global config
that could be easily modified somewhere. Here we change the behavior of
the export API thru hijacking it with self defined code.

For supporting non-strict test suite, the `strict=False` is explicitly
passed into the export API when it's called w/ or w/o strict arg.

* For existing failed strict test cases, non-strict also fails.
* For passed strict but failed non-strict cases, we mark them as
`testing.expectedFailureNonStrict`.
* Moreover, I manually check the failure reason and some of them are not
related to nn.Module asserting exception. I mark them as `# Need to fix
for non-strict mode`.

Test Plan:
python test/export/test_export_nonstrict.py

[ghstack-poisoned]

andrewlee302 commented:
#pytorchbot merge

andrewlee302 commented:
@pytorchbot merge

atalman commented:
@pytorchbot revert -m "OSSCI oncall, broke CI tests" -c nosignal

andrewlee302 committed:
Update on "[Export] Test non-strict mode on existing test cases"


Summary:
Dynamo test methodology provides a good example to patch various
treaments on the same set of test cases. A pitfall is the global config
that could be easily modified somewhere. Here we change the behavior of
the export API thru hijacking it with self defined code.

For supporting non-strict test suite, the `strict=False` is explicitly
passed into the export API when it's called w/ or w/o strict arg.

* For existing failed strict test cases, non-strict also fails.
* For passed strict but failed non-strict cases, we mark them as
`testing.expectedFailureNonStrict`.
* Moreover, I manually check the failure reason and some of them are not
related to nn.Module asserting exception. I mark them as `# Need to fix
for non-strict mode`.

Test Plan:
python test/export/test_export_nonstrict.py

[ghstack-poisoned]

andrewlee302 committed:
Update on "[Export] Test non-strict mode on existing test cases"


Summary:
Dynamo test methodology provides a good example to patch various
treaments on the same set of test cases. A pitfall is the global config
that could be easily modified somewhere. Here we change the behavior of
the export API thru hijacking it with self defined code.

For supporting non-strict test suite, the `strict=False` is explicitly
passed into the export API when it's called w/ or w/o strict arg.

* For existing failed strict test cases, non-strict also fails.
* For passed strict but failed non-strict cases, we mark them as
`testing.expectedFailureNonStrict`.
* Moreover, I manually check the failure reason and some of them are not
related to nn.Module asserting exception. I mark them as `# Need to fix
for non-strict mode`.

Test Plan:
python test/export/test_export_nonstrict.py

[ghstack-poisoned]

atalman commented:
@pytorchbot revert -m "OSSCI oncall, broke CI tests" -c nosignal
